prepare-incremental-build/action.yml

```diff
@@ -0,0 +1,52 @@
+name: 'Prepare incremental kernel build'
+description: 'Pull cached kernel build output from previous runs and prepare the repo for incremental build'
+inputs:
+  repo-root:
+    description: "Root of the kernel repository"
+    required: true
+    default: ${{ github.workspace }}
+  base-branch:
+    description: "Base branch for cache lookup"
+    required: true
+    default: bpf-next_base
+  arch:
+    required: true
+    type: string
+    description: "Part of cache lookup key"
+  toolchain_full:
+    required: true
+    type: string
+    description: "Part of cache lookup key"
+  kbuild-output:
+    required: true
+    type: string
+    description: "Path to KBUILD_OUTPUT"
+    default: kbuild-output
+
+runs:
+  using: "composite"
+  steps:
+
+    - name: Get commit meta-data for cache lookup
+      id: get-commit-metadata
+      shell: bash
+      working-directory: ${{ inputs.repo-root }}
+      run: ${GITHUB_ACTION_PATH}/get-commit-metadata.sh ${{ inputs.base-branch }}
+
+    - name: Pull recent KBUILD_OUTPUT contents
+      uses: actions/cache@v4
```

Comment:
I am pretty sure we should set the working-directory here too.
I believe we are lucky it works because the caller sets `repo-root` to `github.workspace`, which happens to be the default working directory.

Reply:
My understading is that this particular step invoking `actinons/cache@v4` does not have to be executed from the repository directory. The path to store the result of cache lookup (if it's a hit) is passed via `path` parameter:
```yaml
    - name: Pull recent KBUILD_OUTPUT contents
      uses: actions/cache@v4
      with:
        path: ${{ inputs.kbuild-output }} # <-- THIS
        key: ...
```

And we path is through as an action parameter, although my impression is it never changes for our workflows.

And then the next step (that runs `prepare-incremental-builds.sh`) expects to be executed in the repository root, but also expects KBUILD_OUTPUT env variable to point to the cached build.

```yaml
    - name: Prepare incremental build
      shell: bash
      env:
        KBUILD_OUTPUT: ${{ inputs.kbuild-output }}
      working-directory: ${{ inputs.repo-root }} # and so here we do have to set path to the repository
```

I guess it is confusing that we do not set working-directory in cache step, but do set it for other steps, but the action should work either way.

Reply:
I am not convinced `path` work as you are explaining it. Based on https://github.com/actions/cache?tab=readme-ov-file#inputs

my understanding is that it will extract the "kbuild-output" directory from the cache archive.

side note... we are not caching as part of libbpf/ci (e.g https://github.com/libbpf/ci/actions/caches shows no cache, we may want to also exercise this in the future.).

Reply:
> I am not convinced path work as you are explaining it. Based on https://github.com/actions/cache?tab=readme-ov-file#inputs
> 
> my understanding is that it will extract the "kbuild-output" directory from the cache archive.

Maybe I explained it badly, but I think we both understand it correctly.

The cache action is designed to store things like gradle caches ([docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#input-parameters-for-the-cache-action)), so I am pretty sure its `path` input is used both for uploading to the "cloud" after job has finished, and then restoring to the same `path` later, if lookup key matches.

Although actually, if the path is relative, then working directory might be important. So to be on the safe side, I'll set working-directory to the kernel repo root.

Reply:
> The cache action is designed to store things like gradle caches ([docs](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#input-parameters-for-the-cache-action)), so I am pretty sure its `path` input is used both for uploading to the "cloud" after job has finished, and then restoring to the same `path` later, if lookup key matches.

agreed

> Although actually, if the path is relative, then working directory might be important. So to be on the safe side, I'll set working-directory to the kernel repo root.

I think this would make the composite action work as expected *if* the kernel-root does not happen to be the github.workspace.

On the other hand, kbuild-output could be anywhere on the filesystem and we should be able to build by providing the absolute path. If we remove the need for kbuild-output to be in kernel-root, we could leverage the cache for workflows outside of kernel-patches/bpf .But that should probably go in a separate PR.

Reply:
It turns out that it is not possible to specify `working-directory` in a step that uses an action (as opposed to running a script). If you try, [job fails on execution](https://github.com/theihor/libbpf-ci/actions/runs/10893669399). So, in our case, passing an absolute `path` to `actions/cache` is the only good option really.